src/NexusMods.App.UI/WorkspaceSystem/GridUtils.cs

```diff
@@ -63,23 +63,31 @@ internal static bool IsPerfectGrid(WorkspaceGridState gridState, bool doThrow =
         var count = currentState.Count;
         Debug.Assert(count > 0);
 
-        var res = currentState.Clear().Add(new PanelGridState(currentState[0].Id, MathUtils.One));
+        var fake = currentState.Clear().Add(new PanelGridState(MakeId(-1), MathUtils.One));
         for (var i = 1; i < count; i++)
         {
-            var possibleStates = GetPossibleStates(res, maxColumns, maxRows);
+            var possibleStates = GetPossibleStates(fake, maxColumns, maxRows);
             var state = possibleStates.First();
 
-            var newPanels = state.Select(x =>
+            var updatedPanels = state.Select(x =>
             {
-                if (x.Id == PanelId.DefaultValue)
-                    x.Id = currentState[i].Id;
+                if (x.Id == PanelId.DefaultValue) x.Id = MakeId(i);
                 return x;
             }).ToArray();
 
-            res = res.UnionById(newPanels);
+            fake = fake.UnionById(updatedPanels);
         }
 
+        var panels = GC.AllocateUninitializedArray<PanelGridState>(length: count);
+        for (var i = 0; i < count; i++)
+        {
+            panels[i] = new PanelGridState(currentState[i].Id, fake[i].Rect);
+        }
+
+        var res = currentState.UnionById(panels);
         return res;
+
+        static PanelId MakeId(int i) => PanelId.From(new Guid(i, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0));
```

Comment:
I'm not really sure what is happening here with the fake thing. Could you add some comments or change some var names to make what is happening clearer?

Reply:
The code creates a "fake" WorkspaceGrid and fills it up with N panels. We then update the current WorkspaceGrid with those fake panel bounds. This resets the sizes and positions of the panels while keeping the same IDs.

Reply:
Ah, clever, thanks for adding the comments.